### PR TITLE
fix promotion in bidiagonal backslash, At_ldiv_B, and Ac_ldiv_B. add missing bidiagonal matrix conversions.

### PR DESCRIPTION
**Pull request by Sacha0**
This snippet

``` julia
m = 10
intvec = fill(1, m);
lbdintmat = Bidiagonal(fill(2, m), fill(1, m-1), false);
(\)(lbdintmat, intvec)
```

yields

``` julia
ERROR: InexactError()
 in naivesub! at linalg/bidiag.jl:253
 [inlined code] from linalg/bidiag.jl:206
 in \ at linalg/bidiag.jl:267
 in eval at ./boot.jl:265
```

due to promotion in the relevant backslash definition not accounting for usually-necessary divisions in forward/back substitution. This pull request fixes the promotion.

Needs tests. Question on that front. The linalg/bidiag tests don't catch this by nature of testing only floating-point types. Two potential approaches:
1. Minimal modification. Add targeted tests for this issue.
2. Significant modification. Expand the set and combinations of types tested.

? Thanks, and best!

### CODE REVIEW

**Comment by Sacha0** [comment]
Seems Appveyor timed out on x86_64 here as well?

**Comment by Sacha0** [comment]
Expanding the linalg/bidiag tests to integer types revealed the same issue with `A(t|c)_ldiv_B` for integer bidiagonal `A` and integer matrix `B`: The last two calls in this snippet
```julia
m = 10
intmat = fill(1, m, m);
lbdintmat = Bidiagonal(fill(2, m), fill(1, m-1), false);
At_ldiv_B(lbdintmat, intmat)
Ac_ldiv_B(lbdintmat, intmat)
```
both fail as above. The offending lines are https://github.com/JuliaLang/julia/blob/25c3659d6cec2ebf6e6c7d16b03adac76a47b42a/base/linalg/bidiag.jl#L240-L241. The same issue exists with a triangular matrix in place of the bidiagonal, but not where `B` is a `StridedVecOrMat` due to shadowing by more specialized methods in base/linalg/triangular.jl. Sparse `B` demonstrates the issue though:
```julia
m = 10
intmat = fill(1, m, m);
ltintmat = LowerTriangular(rand(1:10, m, m));
At_ldiv_B(ltintmat, sparse(intmat))
Ac_ldiv_B(ltintmat, sparse(intmat))
```
So the two lines linked above need fixing as well.

The linalg/triangular tests fail to catch this issue for triangular matrices as the inner loop over `eltypB` (https://github.com/JuliaLang/julia/blob/25c3659d6cec2ebf6e6c7d16b03adac76a47b42a/test/linalg/triangular.jl#L299) contains only real and complex floating-point types, and they test only `StridedVecOrMat` `B`s in any case. That statement holds for the expanded tests in #14508 as well. So on top of #14508, perhaps the `eltypB` loop in test/linalg/triangular.jl should expand to some integer types, and a test specific to this issue should go into the sparse tests.

I will fix the offending lines and update this PR.

**Comment by Sacha0** [comment]
Should be ready for review. A few points:

The promotion fixes and expanded tests require the added `convert `methods.

The promotion issue involves a generic linear algebra subtlety, described below. Due to that subtlety, the fix for the triangular case is more extensive and largely orthogonal to this PR. So instead of pursuing that fix here, I decoupled the two cases and added a fix-me note to the triangular code. If the approach to the fix in this PR / described below receives approval, I will open a separate issue/PR for the triangular case.

The generic linear algebra subtlety, closely related to recent discussion in #14490: To fix the promotion issue's bidiagonal case, I replaced
```julia
Ac_ldiv_B(A::Union{Bidiagonal, AbstractTriangular}, B::AbstractMatrix) = Ac_ldiv_B!(A,copy(B))
At_ldiv_B(A::Union{Bidiagonal, AbstractTriangular}, B::AbstractMatrix) = At_ldiv_B!(A,copy(B))
```
and the related `\` method, with
```julia
for (f,g) in ((:\, :A_ldiv_B!), (:At_ldiv_B, :At_ldiv_B!), (:Ac_ldiv_B, :Ac_ldiv_B!))
    @eval begin
        function ($f){TA<:Number,TB<:Number}(A::Bidiagonal{TA}, B::AbstractVecOrMat{TB})
            TAB = typeof((zero(TA)*zero(TB) + zero(TA)*zero(TB))/one(TA))
            ($g)(convert(AbstractArray{TAB}, A), copy_oftype(B, TAB))
        end
        ($f)(A::Bidiagonal, B::AbstractVecOrMat) = ($g)(A, copy(B))
    end
end
```
These definitions differ slightly from the related generic promotion method definitions in base/linalg/triangular.jl: I qualified `TA` and `TB` in the promotion methods with `<:Number` and added more general fallback methods. Reason: The generic promotion methods in base/linalg/triangular.jl will fail for matrices with element type `T` lacking `zero(T)`/`one(T)` methods, for example block-matrix constructions; this problem necessitates a more general fallback method. The `<:Number` qualification prevents this more general fallback method from colliding with the generic promotion method; `<:Number` seems the most general qualification that makes sense(?), though this is a great use-case for traits (dispatch on `hasone(T)`, `haszero(T)` or so).

If this seems cogent, I will open an issue/PR to carry these changes to base/linalg/triangular.jl. Thanks, and best!

File: base/linalg/bidiag.jl [hunk 1]
```diff
@@ -67,6 +67,10 @@ function convert{T}(::Type{Tridiagonal{T}}, A::Bidiagonal)
 end
 promote_rule{T,S}(::Type{Tridiagonal{T}}, ::Type{Bidiagonal{S}})=Tridiagonal{promote_type(T,S)}
 
+# Convert from Bidiagonal{Told} to Bidiagonal{Tnew} or AbstractMatrix{Tnew}(->Bidiagonal{Tnew})
+convert{Tnew,Told}(::Type{Bidiagonal{Tnew}}, A::Bidiagonal{Told}) = Bidiagonal(convert(Vector{Tnew}, A.dv), convert(Vector{Tnew}, A.ev), A.isupper)
+convert{Tnew,Told}(::Type{AbstractMatrix{Tnew}}, A::Bidiagonal{Told}) = convert(Bidiagonal{Tnew}, A)
```
Comment (tkelman) [inline base/linalg/bidiag.jl]: is this backwards?

Comment (Sacha0) [inline base/linalg/bidiag.jl]: Which part? Those lines follow https://github.com/JuliaLang/julia/blob/9bab4da7dbee7631319e37a01f039f1d059e3eff/base/linalg/triangular.jl#L23-L24 which seem the same? Also, tests pass which are dependent on the functionality?

Comment (Sacha0) [inline base/linalg/bidiag.jl]: Oh, do you mean the comment? Is the AbstractMatrix{Tnew}(->Bidiagonal{Tnew}) part unclear?

Comment (tkelman) [inline base/linalg/bidiag.jl]: maybe separate the 2 comments and use words for what the -> part is indicating

Comment (Sacha0) [inline base/linalg/bidiag.jl]: Shall I just nix the comment altogether? Those lines seem more or less self-explanatory. Or do you prefer the comment? If the latter, I'll break it in two. Thanks!

Comment (tkelman) [inline base/linalg/bidiag.jl]: the first line is mostly clear, converting the data fields, but the second line isn't quite so obvious. converting to an abstract type is a little bit ambiguous

Comment (andreasnoack) [inline base/linalg/bidiag.jl]: We use these kinds of conversions for element type promotion for all the special matrices. Explaining this in a comment would probably be useful.

File: base/linalg/triangular.jl [hunk 2]
```diff
@@ -1038,6 +1038,11 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
         end
     end
 end
+### Fallbacks brought in from linalg/bidiag.jl while fixing #14506.
+# Eventually the above promotion methods should be generalized as
+# was done for bidiagonal matrices in #14506.
+At_ldiv_B(A::AbstractTriangular, B::AbstractVecOrMat) = At_ldiv_B!(A, copy(B))
+Ac_ldiv_B(A::AbstractTriangular, B::AbstractVecOrMat) = Ac_ldiv_B!(A, copy(B))
```
Comment (andreasnoack) [inline base/linalg/triangular.jl]: Why are these necessary? Shouldn't they be covered by the definitions in https://github.com/JuliaLang/julia/blob/9bab4da7dbee7631319e37a01f039f1d059e3eff/base/linalg/triangular.jl#L996-L1013?

Comment (Sacha0) [inline base/linalg/triangular.jl]: See the last section of https://github.com/JuliaLang/julia/pull/14506#issuecomment-168245349 (that beginning with 'the generic linear algebra subtlety'). Happy new year! :)

File: base/linalg/bidiag.jl (hunk 1 above)
Comment (Sacha0) [inline base/linalg/bidiag.jl]: Revised accordingly. Thanks!

File: base/linalg/triangular.jl (hunk 2 above)
Comment (andreasnoack) [inline base/linalg/triangular.jl]: I don't get it. That section is explaining the methods for `Bidiagonal`, but I don't see an explanation of these methods. I think that the `AbstractTriangular` part of the old definitions might have been redundant because of the definitions I link to above.

Comment (Sacha0) [inline base/linalg/triangular.jl]: I think I see the disconnect. Each generic promotion method in the section you linked contains a line like
```julia
TAB = typeof((zero(TA)*zero(TB) + zero(TA)*zero(TB))/one(TA))
```
or without the `one` in the non-quotient cases. So each such method requires definition of `zero`, and in the quotient cases also `one`, for `T`. But `T` does not necessarily have `one` and/or `zero` methods, for example where `A` is a matrix of matrices. Hence the need for more generic fallbacks like the two lines in question. Does that clarify? Thanks for bearing with me!

Comment (andreasnoack) [inline base/linalg/triangular.jl]: Okay, I get your reasoning now, but, in contrast to your definitions for `Bidiagonal`, the methods I link to don't have restrictions on their element types, so I think the definitions here are shadowed.

Comment (Sacha0) [inline base/linalg/triangular.jl]: Tangentially, I think the two methods in question for `AbstractTriangular` actually cover what I refer to in the immediately preceding comment ('Eventually...'). So perhaps I should nix that comment.

Comment (Sacha0) [inline base/linalg/triangular.jl]: Cheers, and agreed :). Hence the comment immediately preceding the two methods in question: I think the methods you linked require the `<:Number` restriction, at least for now. (In the interim I had forgotten the purpose of that comment --- thanks for reminding me!) But I held off with those modifications to the triangular code till the approach received approval in the bidiagonal case. Thanks again!

**Comment by Sacha0** [comment]
Is this in shape to merge? If so I will open an issue to track the triangular case. Fixing the triangular case appears more involved than fixing the bidiagonal case, and might be better to address after #14005 (which reminds me that I need clean up #14447.) Thanks, and best!

**Comment by andreasnoack** [comment]
Did you check if the methods we discuss were shadowed?

**Comment by Sacha0** [comment]
> Did you check if the methods we discuss were shadowed?

The general promotion methods in linalg/triangular.jl only shadow [the two fallbacks moved from linalg/bidiag.jl] for `StridedVecOrMat` right-hand-sides. To illustrate, the fallbacks handle
```julia
ltA = LowerTriangular(eye(5) + randn(5, 5)/10)
svb = sparsevec([1], [1.0], 5)
At_ldiv_B(ltA, svb)
```
Hence I believe the least dangerous / potentially disruptive approach is to leave those fallbacks in place till the triangular case is fixed properly. Thoughts? Thanks, and best!

File: base/linalg/bidiag.jl [hunk 3]
```diff
@@ -262,9 +265,15 @@ function naivesub!{T}(A::Bidiagonal{T}, b::AbstractVector, x::AbstractVector = b
...
-    TS = typeof(zero(T)*zero(S) + zero(T)*zero(S))
-    TS == S ? A_ldiv_B!(A, copy(B)) : A_ldiv_B!(A, convert(AbstractArray{TS}, B))
+### Generic promotion methods and fallbacks
+for (f,g) in ((:\, :A_ldiv_B!), (:At_ldiv_B, :At_ldiv_B!), (:Ac_ldiv_B, :Ac_ldiv_B!))
+    @eval begin
+        function ($f){TA<:Number,TB<:Number}(A::Bidiagonal{TA}, B::AbstractVecOrMat{TB})
+            TAB = typeof((zero(TA)*zero(TB) + zero(TA)*zero(TB))/one(TA))
+            ($g)(convert(AbstractArray{TAB}, A), copy_oftype(B, TAB))
+        end
+        ($f)(A::Bidiagonal, B::AbstractVecOrMat) = ($g)(A, copy(B))
```
Comment (andreasnoack) [inline base/linalg/bidiag.jl]: What if `eltype(B)` is not a `Number` and not able to store the solution? On the top of my head, that could be something like `eltype(A)=Matrix{Float64}` and `eltype(B)=Matrix{Int}`

Comment (Sacha0) [inline base/linalg/bidiag.jl]: Then the fallback fails, as would the promotion method were the `Number` restriction relaxed. Graceful handling of block matrices seems a broader issue. I've been thinking about that issue in the background following #14490 and hope to partially address it, but I am not prepared to begin doing so here; rather, at some point I will begin with a followup to #14490 and #14493. But perhaps I miss a simple and graceful patch? Thanks again!

Comment (andreasnoack) [inline base/linalg/bidiag.jl]: Okay. That's fine. This still covers move cases than it used to.

**Comment by Sacha0** [comment]
Thanks for the thoughtful review and merge!
